sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala

```diff
@@ -130,16 +130,17 @@ private[r] object SQLUtils {
   }
 
   def dfToCols(df: DataFrame): Array[Array[Any]] = {
-    // localDF is Array[Row]
-    val localDF = df.collect()
+    val localDF: Array[Row] = df.collect()
     val numCols = df.columns.length
+    val numRows = localDF.length
 
-    // result is Array[Array[Any]]
-    (0 until numCols).map { colIdx =>
-      localDF.map { row =>
-        row(colIdx)
-      }
-    }.toArray
+    val colArray = new Array[Array[Any]](numCols)
+    for (colNo <- 0 to (numCols - 1)) colArray.update(colNo, new Array[Any](numRows))
```

Comment:
What about the following code:
    for (colIdx <- 0 until numCols) {
      colArray(colIdx) = new Array[Any](numRows)
    }
    for (rowIdx <- 0 until numRows; colIdx <- 0 until numCols) {
      colArray(colIdx)(rowIdx) = localDF(rowIdx)(colIdx)
    }